### ISSUE TITLE
stdlib.h file not found

### ISSUE THREAD

**Issue by HashingTerry** [nvim-treesitter/nvim-treesitter#2082]
### Before reporting

- [X] I have updated my neovim version to latest _master_
- [X] I have updated my plugin to the latest version
- [X] I have run `:TSUpdate`
- [X] I have read the [troubleshooting section](https://github.com/nvim-treesitter/nvim-treesitter#troubleshooting)

### Describe the bug

nvim treesitter throws an error saying stdlib.h is not found. Occured during TSUpdate. Tried searching on the internet, but sources offer no valid solution. Another occasion where this bug is happening is when I'm fixing the bug ".so is not a valid win32 application", specifically the python parser. By looking at the error I think I'm supposed to install VSCode? The thing is I already have it installed, so I'm clueless. Syntax highlighting only works on LUA and VIM parser as of now (on my machine).
![Screenshot 2021-12-04 110922](https://user-images.githubusercontent.com/87745047/144696932-36ff7d99-ed23-45b8-859f-ec0fda4e1a77.png)

![Screenshot 2021-12-04 111824](https://user-images.githubusercontent.com/87745047/144697010-da8f7fb2-c107-4b73-8413-e732b2e95bf0.png)

### To Reproduce

do TSUpdate

### Expected behavior

No errors

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found  0.20.1 (#) (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v16.13.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `clang` executable found. Selected from { "clang" }
    Version: clang version 13.0.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - lua            ✓ ✓ ✓ ✓ ✓ 
  - vim            ✓ ✓ . . ✓ 
  - python         x x x x x 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}

## The following errors have been detected:
  - ERROR: python(highlights): Failed to load parser: uv_dlopen: C:\Users\####\AppData\Local\nvim\plugged\nvim-treesitter\parser\python.so is not a valid Win32 application.

  - ERROR: python(locals): Failed to load parser: uv_dlopen: C:\Users\####\AppData\Local\nvim\plugged\nvim-treesitter\parser\python.so is not a valid Win32 application.

  - ERROR: python(folds): Failed to load parser: uv_dlopen: C:\Users\####\AppData\Local\nvim\plugged\nvim-treesitter\parser\python.so is not a valid Win32 application.

  - ERROR: python(indents): Failed to load parser: uv_dlopen: C:\Users\####\AppData\Local\nvim\plugged\nvim-treesitter\parser\python.so is not a valid Win32 application.

  - ERROR: python(injections): Failed to load parser: uv_dlopen: C:\Users\####\AppData\Local\nvim\plugged\nvim-treesitter\parser\python.so is not a valid Win32 application.
```

### Output of `nvim --version`

```text
NVIM v0.6.0
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compiled by runneradmin@fv-az152-703

Features: -acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM\sysinit.vim"
  fall-back for $VIM: "C:/Program Files/nvim/share/nvim"

Run :checkhealth for more info
```

**Comment by theHamsta** [comment]
The error is `MSVC not found`. It suggest running clang from a developer command promt (e.g. Visual Studio Native Console 64bit). Please check your compiler installation.

**Comment by HashingTerry** [comment]
so should i install msvc? or just checking compiler should do it?

**Comment by theHamsta** [comment]
I think clang should also be working without msvc headers when working in gcc-like mode. I don't use Windows. Can you try out one of the installation strategies discribed in the README?

**Comment by HashingTerry** [comment]
ok thanks

**Comment by theHamsta** [comment]
@HashingTerry sorry, meant to say wiki. In the Wiki, people help each out with tips on how to make it run on windows

**Comment by HashingTerry** [comment]
oh ok thanks

**Comment by HashingTerry** [comment]
so, I have tried installing MSVC, also I reinstalled clang, rebooted my computer and it's not working. Perhaps maybe I did something wrong with installing clang?

**Comment by HashingTerry** [comment]
@theHamsta I was able to compile it now as I switched to gcc for the compiler. But it throws another error saying "so is not a valid win32 app", i figured I could try use precompiled parsers. Can you send link to precompiled parsers?

**Comment by theHamsta** [comment]
https://github.com/nvim-treesitter/nvim-treesitter/actions/runs/1531376235 you can get there by clicking on the tick next to a commit 

![image](https://user-images.githubusercontent.com/7189118/144710174-40ea40e1-39d2-42a0-b996-02e95d4e8bf9.png)
and then selectinging `Summary`

![image](https://user-images.githubusercontent.com/7189118/144710198-7a2f0d4b-3b30-46f4-a3c5-d720ae03182e.png)


Have you tried using MSVC or clang from this terminal: https://stackoverflow.com/questions/61209155/how-do-i-get-the-x64-native-tools-developer-command-prompt-for-visual-studio-com

**Comment by HashingTerry** [comment]
nvim is 64 bit
gcc/mingw is 32 bit

**Comment by theHamsta** [comment]
You must compile a 64bit parser for a 64bit nvim. Otherwise it won't work.

**Comment by HashingTerry** [comment]
oh ok

**Comment by HashingTerry** [comment]
do you know any reliable source to download 64 bit mingw?

**Comment by theHamsta** [comment]
@HashingTerry would you mind sharing where you got your mingw? Maybe even in the wiki?

**Comment by HashingTerry** [comment]
https://techdecodetutorials.com/download/

Make sure to select the windows 10 version.

**Comment by ulisses-cruz** [comment]
I'm having the same problem but I'm on Mac.

I see this error when I run `:TSUpdate`

```zsh
nvim-treesitter[lua]: Error during compilation
In file included from src/parser.c:1:
./src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
#include <stdlib.h>
         ^~~~~~~~~~
1 error generated.
In file included from src/scanner.c:1:
./src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
#include <stdlib.h>
         ^~~~~~~~~~
1 error generated.
```

**Comment by ShortArrow** [comment]
## Summery

I have same problem.
When run `:TSUpdate`

## Error messages
```bash
[nvim-treesitter] [8/10, failed: 8] Checking out locked revision
[nvim-treesitter] [8/10, failed: 8] Compiling...nvim-treesitter[tsx]: Error during compilation
clang: warning: unable to find a Visual Studio installation; try running Clang from a developer command prompt [-Wmsvc-not-found]^M
In file included from src/parser.c:1:^M
./src\tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found^M
#include <stdlib.h>^M
         ^~~~~~~~~~^M
1 error generated.^M
In file included from src/scanner.c:1:^M
In file included from ./src/../../common/scanner.h:1:^M
./src\tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found^M
#include <stdlib.h>^M
         ^~~~~~~~~~^M
1 error generated.^M
```

## checkhealth

```bash
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
   - OK: `tree-sitter` found 0.20.7 (b268e412ad4848380166af153300464e5a1cf83f) (parser generator, only needed for :TSInstallFromGrammar)
   - OK: `node` found v18.12.1 (only needed for :TSInstallFromGrammar)
   - OK: `git` executable found.
   - OK: `clang` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: clang version 15.0.1
   - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## OS Info:
{
   machine = "x86_64",
   release = "10.0.22621",
   sysname = "Windows_NT",
   version = "Windows 10 Home"
}
 
## Parser/Features H L F I J
 
   Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
          +) multiple parsers found, only one will be used
          x) errors found in the query, try to run :TSUpdate {lang}
```

## versions
```powershell
> & 'C:\Program Files\LLVM\bin\clang.exe' --version
clang version 15.0.5
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin

> clang --version
clang version 15.0.1
Target: i686-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\bin

> nvim --version
NVIM v0.8.0-1210-gd367ed9b2
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compiled by runneradmin@fv-az177-603

Features: -acl +iconv +tui
See ":help feature-compile"

      システム vimrc: "$VIM\sysinit.vim"
       省略時の $VIM: "C:/Program Files (x86)/nvim/share/nvim"

Run :checkhealth for more info
```

## chocolatey

```powershell
> choco search clang
Chocolatey v1.1.0
llvm 15.0.5 [Approved] Downloads cached for licensed users
unreal-linux-toolchain 20.13.0.1
winlibs 10.11.8.0 [Approved] Downloads cached for licensed users
codeblocks 20.3.0.20210213 [Approved]
4 packages found.

> choco search llvm
Chocolatey v1.1.0
llvm 15.0.5 [Approved] Downloads cached for licensed users
winlibs 10.11.8.0 [Approved] Downloads cached for licensed users
winlibs-llvm-free 10.11.0 [Approved] Downloads cached for licensed users
ldc 1.30.0 [Approved] Downloads cached for licensed users
dotnetcoresdk 1.0.1 [Approved] Downloads cached for licensed users - Possibly broken for FOSS users (due to original download location changes by vendor)
ccls 0.20220729 [Approved]
ghc 9.4.3 [Approved]
7 packages found.
```

## Visual Studio Installer.exe --locale en-US
![image](https://user-images.githubusercontent.com/16986253/208393297-173d2526-0aac-45f0-9287-e26d7ecd0e3f.png)

**Comment by theHamsta** [comment]
@ShortArrow are you running from a VS native tool command prompt? https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170 the microsoft tooling sometimes needs some special environments variables to be set

**Comment by ShortArrow** [comment]
@theHamsta 
I am on the Normal Powershell 7 Latest. I was setted clang the in PATH of System Enviroment, and rebooted every check. And priority order changing of clang from choco or from MSInstaller, it was no effects .

After your comment, I tried MSVC Installing.

![image](https://user-images.githubusercontent.com/16986253/208433857-e5b1344b-f382-4664-ba26-9d3c3dd4d65f.png)

The error message `clang: warning: unable to find a Visual Studio installation; try running Clang from a developer command prompt [-Wmsvc-not-found]`  was erased. But new error messages are comming.

```bash
  tsx(highlights) is concatenated from the following files:
  | [ERROR]:"C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\queries\tsx\highlights.scm", 
  failed to load: Failed to load parser: uv_dlopen: C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\parser\tsx.so is not a valid Win32 application.
- ERROR: tsx(locals): Failed to load parser: uv_dlopen: C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\parser\tsx.so is not a valid Win32 application.
  tsx(locals) is concatenated from the following files:
  | [ERROR]:"C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\queries\tsx\locals.scm", 
  failed to load: Failed to load parser: uv_dlopen: C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\parser\tsx.so is not a valid Win32 application.
- ERROR: tsx(folds): Failed to load parser: uv_dlopen: C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\parser\tsx.so is not a valid Win32 application.
  tsx(folds) is concatenated from the following files:
  | [ERROR]:"C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\queries\ecma\folds.scm",
  failed to load: Failed to load parser: uv_dlopen: C:\Users\who\AppData\Local\nvim-data\site\pack\packer\start\nvim-treesitter\parser\tsx.so is not a valid Win32 application.
```

Found ` is not a valid Win32 application.`.
So, As mentioned in the comment above, it is thought that it may be necessary to align with 64bit or 32bit. So I will try.

**Comment by marvintensuan** [comment]
Hey guys. Here's what worked for me.

### My setup

- Neovim installed via winget
- I already had MSVC installed. I subsequently installed llvm/clang via Chocolatey.
- Windows Terminal (Run as Administrator)
- Windows 10, 64-bit

### Things I did...

... in chronological order.  Took me a day so I might have missed out a few details.

1. Upgrade to MSVC v143. Gave up eventually and installed llvm/clang via chocolatey.
2. [Getting Started with the LLVM System using Microsoft Visual Studio](https://llvm.org/docs/GettingStartedVS.html) - these lines only since I've already downloaded [1]:
    
```
    regsvr32 "%VSINSTALLDIR%\DIA SDK\bin\msdia140.dll"
    regsvr32 "%VSINSTALLDIR%\DIA SDK\bin\amd64\msdia140.dll"
```

3. Install Windows Universal CRT SDK (available in Visual Studio Installer). [2]
4. Install Windows 10 (or [11](https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/)) SDK (available in Visual Studio Installer). [3]

Footnotes:

[1] Might wanna `echo %VSINSTALLDIR%`. Best to do everything in _Developer Command Prompt for VS 2022_. Run these lines every time you open to ensure environment is correctly set up:

- "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\Tools\VsDevCmd.bat"
- "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat"

[2] I'm no C nor C++ guy but after having two compilers in my PC, it astounds me how both have no explicit `stdlib.h` file in their respective `include/` directories. This step apparently solves this.

[3] This step solved this error:  Error LNK1104: cannot open file 'kernel32.lib'.

**Comment by ShortArrow** [comment]
After all. I got health of TS in [this way](https://github.com/osherbz/nvim-treesitter/wiki/Windows-support#troubleshooting) of @osherbz 's repository.

When I was run `:TSInstall` , neovim or llvm was 32bit and opposite one was 64bit. So my treesitter was dirty. 

I was reinstall neovim and llvm and clang and zig. But only `:TSUpdate` or only `:TSInstall` like [this way](https://github.com/nvim-treesitter/nvim-treesitter/wiki/Windows-support#troubleshooting) were no effects to get clean on my environment.

I have using packer. So I needed these steps, for reinstall nvim-treesitter.

1. Delete nvim-treesitter from dependency of my neovim config.
2. Reopen neovim and run Excommand `:PackerSync`.
3. Add nvim-treesitter to dependency of my neovim config.
4. Reopen neovim and run Excommand `:PackerSync`.
5. `checkhealth nvim-treesitter` has get clean.

Check the layer, like bellow.

- Health of installed nvim
- Health of llvm or msvc or zig
- Health of installed nvim-treesitter
- Health of installed modlues of treesitter
- Same architecture 64 or 32

Thank you 😄

**Comment by xy777tor** [comment]
Despite you all have probably solve this problem already I decided to type some info about this issue.
I suffered with this problem for several days and it was because of I had "Free Pascal" on my computer (I studied it a couple years ago). "Free Pascal" was in PATH variable and it had gcc.exe file in it's folder. This version was too old I think. The solution may be that you have different versions of the compiler.

**Comment by wwwggxk** [comment]
> I'm having the same problem but I'm on Mac.
> 
> I see this error when I run `:TSUpdate`
> 
> ```shell
> nvim-treesitter[lua]: Error during compilation
> In file included from src/parser.c:1:
> ./src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
> #include <stdlib.h>
>          ^~~~~~~~~~
> 1 error generated.
> In file included from src/scanner.c:1:
> ./src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
> #include <stdlib.h>
>          ^~~~~~~~~~
> 1 error generated.
> ```

I also have the same problem.
1. ":checkhealth nvim-treesitter" show:
```
Installation
  5 - OK tree-sitter found 0.20.8 (parser generator, only needed for :TSInstallFromGrammar)
  6 - OK node found v16.14.2 (only needed for :TSInstallFromGrammar)
  7 - OK git executable found.
  8 - OK /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang executable found. Selected from { "/Applications/
    Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang", "cc", "gcc", "clang", "cl", "zig" }
  9   Version: Apple clang version 15.0.0 (clang-1500.0.40.1)
 10 - OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.
```

2. But not found c library in the directory:
<img width="771" alt="image" src="https://github.com/nvim-treesitter/nvim-treesitter/assets/7045546/77a8386a-ce40-4bad-bbff-e2382d4273bf">

The c library is located at:
<img width="524" alt="image" src="https://github.com/nvim-treesitter/nvim-treesitter/assets/7045546/ab9254ac-fc25-45aa-acc6-52fa60d41341">

3. So just add the "export CPATH=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/
" to .zshrc file, then it works;

**Comment by patrickjoan** [comment]
> I'm having the same problem but I'm on Mac.
> 
> I see this error when I run `:TSUpdate`
> 
> ```shell
> nvim-treesitter[lua]: Error during compilation
> In file included from src/parser.c:1:
> ./src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
> #include <stdlib.h>
>          ^~~~~~~~~~
> 1 error generated.
> In file included from src/scanner.c:1:
> ./src/tree_sitter/parser.h:10:10: fatal error: 'stdlib.h' file not found
> #include <stdlib.h>
>          ^~~~~~~~~~
> 1 error generated.
> ```

This a lot of time later but i ran into the same issue on a mac. For me what solved it was using
"sudo xcode-select --reset" After that it worked.

Hope this helps if someone has the same error.

**Comment by MrOxMasTer** [comment]
After reading a lot of information for a long time, I came to a conclusion on this topic.

The author requires for his library: `compiler C and libstdc++ `.  A C compiler and a library special for C++.

I do not know why the author indicated that llvm is suitable here, provided that gcc (MinGW) has the necessary library for C++, namely libstdc++. There is a library in llvm: libc++

![image](https://github.com/user-attachments/assets/b117500d-8bea-4769-9086-4c5cbc174a26)

I took this information from the MSYS2 website. It is not directly related, but as we see naturally, there is one library for C++ in llvm, and the one we need for gcc

To put it simply: `install gcc so as not to sweat`

There was also a problem with MSVC, but as I understand it, it was solved by installing the compiler from VS

No additional components installed in VS starting with MSVC and others, no "prompt/powershell special/native", helped me to cope with this problem. The errors went away, but the fact that there was no library is obvious.

Perhaps it was necessary to install this very library for clang (it belongs to llvm), but how and by a miracle, I do not know

**Comment by mehdiAI** [comment]
When i install treesitter in nvim and set treesitter.lua file to install some languages in ensure_installed variable,
but i got c++ compiler and stdlib.h not found errors, then i resolved it by below approach:

Install clang for c++ compiler and after that need to visual cpp build tool.
Links:
https://github.com/llvm/llvm-project/releases/tag/llvmorg-18.1.8
https://visualstudio.microsoft.com/visual-cpp-build-tools/
